### PR TITLE
First implementation of gas clustering

### PR DESCRIPTION
**Pull request by AmeliaNadal**
This branch is suppose to allow the clustering of the gas grid : CH4, H2_grid and H2_ind_load buses, related links, generators, loads and stores.

Important remarks:
- The gas clustering should append after the clustering of the other new sectors or more precisely after the clustering of the H2_saltcaverns buses. Therefore, the branches [features/356_assign_country_to_buses ](https://github.com/openego/eTraGo/tree/features/356_assign_country_to_buses) and [features/#333_spatial_clustering_for_sector_coupling](https://github.com/openego/eTraGo/tree/features/%23333_spatial_clustering_for_sector_coupling) where merged into this branch.
- This branch has been tested only on a reduce dataset (test mode of egon-data) and without the CHP technologies.
- This clustering is functionning only for the eGon2035 scenario datasets, this should become more generic: see https://github.com/openego/eTraGo/issues/358

### CODE REVIEW

File: etrago/cluster/gasclustering.py
Comment (ClaraBuettner) [inline etrago/cluster/gasclustering.py:115]: Other tables are imported in etrago.io. If this table is needed we have to make sure that it is part of the eTraGo input database (and e.g. also publish it).

File: etrago/cluster/gasclustering.py
Comment (ClaraBuettner) [inline etrago/cluster/gasclustering.py:374]: I think you should use` np.sum` for `p_nom_max` and `p_nom_min` because you want to sum up the capacity limits.

File: etrago/appl.py
Comment (ClaraBuettner) [inline etrago/appl.py:88]: In a new environment, the clustering is not working because the `n_jobs` parameter was removed in scikit learn in  versions `>0.23`. We can either restrict the package version or remove the parameter from our functions. 
I have the tendency to use the second option to avoid new conflicts of package versions. What do you think?
@CarlosEpia: It is the same for clustering the electricity grid, we should agree one one solution

Comment (CarlosEpia) [reply etrago/appl.py:88]: I think the best option is to remove the n_jobs parameter. I will do it in my branch #361

File: etrago/cluster/gasclustering.py
Comment (AmeliaNadal) [reply etrago/cluster/gasclustering.py]: Thanks for the correction, I've fixed this in the the commit https://github.com/openego/eTraGo/pull/359/commits/8b7ea56190e2d833fa9e28783c3c8cd29a0a5057.

**Comment by ClaraBuettner** [comment]
> This branch is suppose to allow the clustering of the gas grid : CH4, H2_grid and H2_ind_load buses, related links, generators, loads and stores.
> 
> Important remarks:
> 
>     * The gas clustering should append after the clustering of the other new sectors or more precisely after the clustering of the H2_saltcaverns buses. Therefore, the branches [features/356_assign_country_to_buses ](https://github.com/openego/eTraGo/tree/features/356_assign_country_to_buses) and [features/#333_spatial_clustering_for_sector_coupling](https://github.com/openego/eTraGo/tree/features/%23333_spatial_clustering_for_sector_coupling) where merged into this branch.
> 
>     * This branch has been tested only on a reduce dataset (test mode of egon-data) and without the CHP technologies.
> 

I tested the gas clustering for Schleswig-Holstein and Germany and it was working as expected. The CHPs are also clustered correctly.

**Comment by CarlosEpia** [comment]
I tested this branch and it is running for me. It looks ok. I have two questions about the methodology:
1. This approach clusters each kind of gas-bus according to the parameter 'n_clusters_gas'. Is that actually what we want?
2. The clustering is not considering the buses in other countries, which means that in the end, we have more buses than the given by the 'n_clusters_gas' parameter. It works differently for the eHV and k-mean clustering for the other sectors. Should we agree on a unique approach regarding this topic?

File: etrago/appl.py
Comment (AmeliaNadal) [reply etrago/appl.py]: This issue will be discussed further here: https://github.com/openego/eTraGo/issues/364.

File: etrago/cluster/gasclustering.py
Comment (ClaraBuettner) [reply etrago/cluster/gasclustering.py]: I think the new option is better, thank you!

File: etrago/cluster/gasclustering.py
Comment (ClaraBuettner) [inline etrago/cluster/gasclustering.py:299]: I'm wondering why you are clustering the links twice (line 286 and line 297). Do you want to cluster links bidirectional?

File: etrago/cluster/gasclustering.py
Comment (ClaraBuettner) [inline etrago/cluster/gasclustering.py:315]: Time-depended values for links are imported, but there are time series values for some links_id which are not in the links dataframe :
```

WARNING:pypsa.io:Components Index(['1740', '1741', '1742', '1743', '1744', '1745', '1746', '1747', '1748',
       '1749',
       ...
       '2005', '2006', '2007', '2008', '2009', '2010', '2011', '2012', '2013',
       '2014'],
      dtype='object', length=216) for attribute efficiency of Link are not in main components dataframe links
```
I will try to find out why this happens.

Comment (ClaraBuettner) [reply etrago/cluster/gasclustering.py:315]: The warnings seem to be related to #366. It could be solved by changing the order of the clustering and clustering the gas network before clustering the electric grid. Is there anything against changing the execution order like this?

Comment (CarlosEpia) [reply etrago/cluster/gasclustering.py:315]: I also think it is related to #366. The clustering order must be irrelevant, so if it helps to swap the clusterings, I am not against it.

Comment (fwitte) [reply etrago/cluster/gasclustering.py:315]: I am working on https://github.com/openego/eTraGo/pull/367, it should be possible to modify the aggragation that in a way that the "in-between" technologies (generalized: carriers that are connected to clustered network structures) are aggregated after the busmaps for gas and AC have been created, indendetly of the clustering order.

File: etrago/cluster/gasclustering.py
Comment (ClaraBuettner) [reply etrago/cluster/gasclustering.py:299]: But there could be problems when you have unidirectional links which should be not combined with links going into the other direction. If you have in the network with the full complexity unidirectional links in both directories, you don't want to combine them and have only one link into one direction. Or am I mistaken?

File: etrago/cluster/gasclustering.py
Comment (ClaraBuettner) [reply etrago/cluster/gasclustering.py:315]: So how do we want to proceed? Should we wait until #367 is finished and merged into this branch before merging it into dev?

Comment (CarlosEpia) [reply etrago/cluster/gasclustering.py:315]: For me, it makes sense to merge #367 into this branch first and then check that everything is working together as @ClaraBuettner suggested.

Comment (ClaraBuettner) [reply etrago/cluster/gasclustering.py:301]: This issue will be solved after #377 is merged into dev.

**Review by ClaraBuettner — APPROVED** [review]

**Comment by AmeliaNadal** [comment]
Closes https://github.com/openego/eTraGo/issues/343

**Pull request merged**
